2215 allow lambda in sympy

hiker opened the pull request:
Fixes #2215.

rupertford commented:
Hi @hiker, this appears to work for my example - whoop. I say "appears to" as there is another error in this example that I've fixed on a separate branch but can't test everything is OK for the example unless I joint the two branches together. However, that error is to do with kernels and yours was to do with the algorithm code so I'm pretty confident you've fixed it.

hiker commented:
Ready for review.

hiker commented:
I saw you assigning it to me - not exactly sure why, I don't see any comments :) FWIW, I've updated it to latest master.

rupertford commented on src/psyclone/psyir/tools/dependency_tools.py:
```diff
@@ -383,12 +383,19 @@ def _get_dependency_distance(var_name, index_read, index_written):
+        # (e.g. if a Fortran variable 'lambda' is used, it will be
+        # renamed to lambda_1, and the type map will have lambda_1 as key
+        # and the SymPy symbol for 'lambda' as value).
+        for var in symbol_map.values():
```
Can't you just try to access it in a try with a KeyError?
```
try:
    result = symbol_map[var_name]
except KeyError:
    ...
```

hiker replied:
No, I hoped that my explanation was good enough: e.g. in the case of `lambda`, the key will be `lambda_1`, but the value `lambda`.  So I would need to use a reverse dictionary to look it up, but I am not sure if this is worth computing.

rupertford commented on src/psyclone/tests/core/symbolic_maths_test.py:
```diff
@@ -356,6 +356,8 @@ def test_symbolic_math_functions_with_constants(fortran_reader, expressions):
 
 
 @pytest.mark.parametrize("expressions", [("field(1+i)", "field(i+1)"),
+                                         ("lambda", "lambda"),
```
Perhaps add to the test description that it could also be a Python Keyword.

hiker replied:
Done.

rupertford commented on src/psyclone/tests/psyir/backend/sympy_writer_test.py:
```diff
@@ -512,3 +512,45 @@ def test_sympy_writer_user_types(fortran_reader, fortran_writer,
+                integer :: x
+                x = {expression}
+                end program test_prog '''
+    psyir = fortran_reader.psyir_from_source(source)
```
This could be shortened if you used `psyir_from_expression` but I won't enforce it as it is already clear what you are doing.

hiker replied:
I think I tried that. Issue was that `psyir_from_expression` needs a symbol table, and that the symbols need to be declared ... which was a lot more work than just parsing a sample code.

rupertford replied:
I'm happy with that. As I said I think the example is already clear.

rupertford commented on src/psyclone/tests/psyir/tools/dependency_tools_test.py:
```diff
@@ -779,3 +779,72 @@ def test_da_array_expression(fortran_reader):
+
+# -----------------------------------------------------------------------------
+def test_reserved_words(fortran_reader):
+    '''Tests that using a reserved Python word ('lambda' here') as loop
```
as a loop

hiker replied:
Done.

rupertford commented on src/psyclone/tests/psyir/tools/dependency_tools_test.py:
```diff
@@ -779,3 +779,72 @@ def test_da_array_expression(fortran_reader):
+def test_reserved_words(fortran_reader):
+    '''Tests that using a reserved Python word ('lambda' here') as loop
+    variable, which will be renamed when converting to SymPy, works as
+    expected. It also makes sure that name clashes are handled as expected by
```
Also makes sure that

hiker replied:
Done.

rupertford commented on src/psyclone/tests/psyir/tools/dependency_tools_test.py:
```diff
@@ -779,3 +779,72 @@ def test_da_array_expression(fortran_reader):
+    assert msg.code == DTCode.ERROR_WRITE_WRITE_RACE
+    assert msg.var_names == ["mask(jk,jk)"]
+
+    # Write to array that does not depend on parallel loop variable
```
please change "parallel loop variable" to "the parallel loop variable" from here on.

hiker replied:
All done.

rupertford reviewed (changes requested):
Nicely done. My comments are all minor.

rupertford commented:
Back to @hiker

hiker commented on src/psyclone/psyir/backend/sympy_writer.py:
```diff
@@ -91,6 +94,10 @@ class SymPyWriter(FortranWriter):
 
+    # A list of all reserved Python keywords (Fortran variables that are the
+    # same as a reserved name must be renamed, otherwise parsing will fail):
+    _RESERVED_NAMES = set()
```
Done.

hiker commented on src/psyclone/psyir/backend/sympy_writer.py:
```diff
@@ -252,34 +297,33 @@ def _create_type_map(self, list_of_expressions):
+        # create guaranteed unique names for lower and upper bounds.
         self._symbol_table = SymbolTable()
+        for reserved in SymPyWriter._RESERVED_NAMES:
+            self._symbol_table.find_or_create(reserved)
```
That's indeed better.

hiker commented on src/psyclone/psyir/backend/sympy_writer.py:
```diff
@@ -252,34 +297,33 @@ def _create_type_map(self, list_of_expressions):
                 name = ref.name
-                if name in self._symbol_table:
-                    # The name has already been declared, ignore it now
+                if name in self._symbol_table.tags_dict:
```
Done.

hiker commented on src/psyclone/psyir/backend/sympy_writer.py:
```diff
@@ -252,34 +297,33 @@ def _create_type_map(self, list_of_expressions):
-                # as done
-                self._symbol_table.find_or_create(name)
-
+                unique_sym = self._symbol_table.new_symbol(name, tag=name)
```
Done.

hiker commented:
OK, I've addressed most issues. I did leave the full-test-program-parsing in (instead of `psyir_from_expression`), and not sure what to do about the dictionary lookup suggestion. Could you try to check the comment just before that, I am not really sure what to add.

rupertford reviewed (approved):
All comments have ben addressed. I've brought the branch up to master and all is green for the local CI and for the remote CI compilation and running. Approving.
